src/cockpit/polkit.py

```diff
@@ -0,0 +1,139 @@
+import asyncio
+import locale
+import logging
+import os
+import pwd
+
+from cockpit._vendor.systemd_ctypes import bus
+
+# that path is valid on at least Debian, Fedora/RHEL, and Arch
+HELPER_PATH = '/usr/lib/polkit-1/polkit-agent-helper-1'
+
+logger = logging.getLogger(__name__)
+
+
+# https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.AuthenticationAgent.html
+class org_freedesktop_PolicyKit1_AuthenticationAgent(bus.Object):
+    def __init__(self, responder):
+        super().__init__()
+        self.responder = responder
+
+    @bus.Interface.Method('', ['s', 's', 's', 'a{ss}', 's', 'a(sa{sv})'])
+    async def begin_authentication(self, action_id, message, icon_name, details, cookie, identities):
+        logger.debug('BeginAuthentication: %s %s %s %s %s %s',
+                     action_id, message, icon_name, details, cookie, identities)
+        # only support authentication as ourselves
+        my_uid = os.geteuid()
+        for (auth_type, subject) in identities:
+            if auth_type == 'unix-user' and 'uid' in subject and subject['uid'].value == my_uid:
+                logger.debug('Authentication subject %s matches our uid %d', subject, my_uid)
+                break
+        else:
+            logger.warning('Not supporting authentication as any of %s', identities)
+            return
+
+        user_name = pwd.getpwuid(my_uid).pw_name
+        process = await asyncio.create_subprocess_exec(HELPER_PATH, user_name, cookie,
+                                                       stdin=asyncio.subprocess.PIPE,
+                                                       stdout=asyncio.subprocess.PIPE)
+        while process.returncode is None:
```

Comment:
I feel like the main "dialog" part of this function ought to be in a separate function.

Reply:
...and then I'd lift out all of the various cases of error and cancellation handling into a single `finally:` block which makes sure that the subprocess is correctly cleaned up.

Reply:
Done.